examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,217 @@
+parser.add_argument('-n_epoch', default=5000, type=int)
+parser.add_argument('-dtype', default='float32', type=str,
+                    choices=['float32', 'float64'])
+parser.add_argument('-dataset', default='boston_housing', type=str,
+                    choices=['boston_housing', 'protein_data'])
+
+
+def rbf(x, y, inv_scale):
+    x = tf.expand_dims(x, 1)
+    y = tf.expand_dims(y, 0)
+    inv_scale = tf.reshape(inv_scale, [1, 1, -1])
+    return tf.exp(- tf.reduce_sum(tf.square(x - y) * inv_scale, axis=-1) / 2)
+
+
+@zs.reuse('model')
+def build_model(observed, x_ph, hps, n_particles, full_cov=False):
+    '''
+    Build the SVGP model.
+    Note that for inference, we only need the diagonal part of Cov[Y], as ELBO
+    equals sum over individual observations.
+    For visualization etc we may want a full covariance. Thus the argument
+    `full_cov`.
+    '''
+    with zs.BayesianNet(observed) as model:
+        z_pos = tf.get_variable(
+            'z/pos', [hps.n_z, hps.n_covariates], hps.dtype,
+            initializer=tf.random_uniform_initializer(-1, 1))
+        k_inv_scale = tf.get_variable(
+            'k_inv_scale', [hps.n_covariates], hps.dtype,
+            tf.zeros_initializer())
+        k_inv_scale = tf.exp(k_inv_scale)
+
+        Kzz = rbf(z_pos, z_pos, k_inv_scale)
+        Kxz = rbf(x_ph, z_pos, k_inv_scale)
+        Kzz_chol = tf.cholesky(Kzz)
+        fZ = zs.MultivariateNormalCholesky(
+            'fZ', tf.zeros([hps.n_z], dtype=hps.dtype), Kzz_chol,
+            n_samples=n_particles)
+
+        # Model Y|Z.
+        # Mean[Y|Z] = Kxz @ inv(Kzz) @ fZ
+        # Cov[Y|Z] = Kxx - Kxz @ inv(Kzz) @ Kzx + noise_level * I
+        noise_level = tf.get_variable(
+            'noise_level', shape=[], dtype=hps.dtype,
+            initializer=tf.constant_initializer(0.5))
+        noise_level = tf.nn.softplus(noise_level)
+
+        # With ill-conditioned Kzz, the inverse is often asymmetric, which
+        # breaks further cholesky decomposition. We compute a symmetric one.
+        Kzz_chol_inv = tf.matrix_triangular_solve(
+            Kzz_chol, tf.eye(hps.n_z, dtype=hps.dtype))
+        Kzz_inv = tf.matmul(tf.transpose(Kzz_chol_inv), Kzz_chol_inv)
+
+        Kxziz = tf.matmul(Kxz, Kzz_inv)
+        mean_y = tf.matmul(fZ, tf.transpose(Kxziz)) # [n_particles, n_cov]
+
+        cov_fx_given_fz = rbf(x_ph, x_ph, k_inv_scale) - \
+            tf.matmul(Kxziz, tf.transpose(Kxz))
+        cov_y = cov_fx_given_fz + noise_level * tf.eye(
+            tf.shape(x_ph)[0], dtype=hps.dtype)
+        if full_cov:
+            cov_y_chol = tf.tile(tf.expand_dims(tf.cholesky(cov_y), 0),
+                                 [n_particles, 1, 1])
+            y = zs.MultivariateNormalCholesky('y', mean_y, cov_y_chol)
+        else:
+            std_y = tf.sqrt(tf.matrix_diag_part(cov_y))
+            std_y = tf.tile(tf.expand_dims(std_y, 0), [n_particles, 1])
+            y = zs.Normal('y', mean=mean_y, std=std_y, group_ndims=1)
```

Comment:
A final question. The else branch seems not right if you don't marginalize over inducing points as the doubly stochastic DGP paper does. y is only independent across datapoints conditioning on f, but not fz.

Reply:
The code follows from the fact that ELBO can be decomposed into sum over observations, $E_{q(f_z)}[\log p(y,f_z) - \log q(f_z)] = \sum_i E_{q(f_z)}[\log p(y_i,f_z)] - E_q[\log q(f_z)]$. And $p(y_i|f_z)$ only depends on the diagonal part of $Cov[f|f_z]$. It seems correct to me.
The code is really misleading though: the model spec should only contains f_y drawn from GP prior, instead of the current form (fz is part of the inference params for our formulation). I'll fix it recently.

Reply:
I spent some time thinking about this and felt it is not easy to fix. Now I have a solution, which is that we add a feature to our vi algorithms: If a variational posterior is not provided for a latent (e.g., `f` here), then the algorithm will automatically choose the prior of this latent as the variational (e.g., `p(f|u)` here). To achieve this the variational objective should require the model function instead of a log_joint, which is a change that has been in my mind for long time and will benefit other things as well.

Reply:
Sorry for the late reply - just finished the exams. It's doable, although it just needs me to copy-paste the `p(f|u)` term from model to variational, without changing the basics. 
The problem is that now the training is `O(N_batch^3)` instead of `O(N_batch^2 N_Z)` as we need to sample from `p(f|u)`, which seems not avoidable if we implement VFE (the approximation the DGP paper cites / I claimed to have implemented) and do not utilize conjugacy.
Alternatively, we can implement FITC (mentioned [here](http://www2.aueb.gr/users/mtitsias/papers/BARK08titsias.pdf) and [here](gpss.cc/gpss17/slides/gp-approx-new.pdf)) which seems to match what I've done, and is faster in the non-conjugate case. The downside is that the standard formulation is not fully Bayesian (we optimize inducing point location instead of sampling from them), and does overfit in practice. What do you think?

Reply:
You mean sampling from \int q(u)p(f|u) du costs O(N_batch^2 N_Z) and sampling from q(u)p(f|u) costs O(N_batch^3)?

Reply:
I thought sampling from q(u)p(f|u) costs O(N_batch^2 N_Z) also.

Reply:
Ah you're right as we only need to sample from the marginal. The ELBO is E[log (p(y|f)p(f|u)p(u)) / (p(f|u)q(u))]. So the only problem is conceptual.

Reply:
> So the only problem is conceptual (why model and variational output marginal distributions).

I don't get it.

> So the current code is correct?

Not really. The ELBO of VFE is
```E_{p(f|u)q(u)} \log [p(y|f)p(f|u)p(u) / p(f|u)q(u)].```
A better practice is (as in the stochastic DGP paper)
```E_{\int p(f|u)q(u) du} \log [p(y|f)p(u) / q(u)]```
In ZhuSuan, if we made the variational objective to be a bit smarter than it currently is (i.e., it will automatically remove the two terms of p(f|u) if no q(f|u) is given, which means you choose to use p(f|u) as q(f|u)), we could do
```E_{p(f|u)q(u)} \log [p(y|f)p(u) / q(u)]```
which is almost the second one with a bit more variance introduced when sampling from q(f) = \int p(f|u)q(u) du. But it is good enough since the complexity is still O(n_batch^2 n_z) and the objective is correct as a VFE.

However, the current implementation is
```E_{q(u)} \log {[p(y|u)p(u) / q(u)] = E_{q(u)} \log {[\int p(y|f)p(f|u) du]p(u) / q(u)]```
which is only correct when no stochastic approximation is used (`p(y|u)` cannot factorize). But you are using mini-batch training, which is not correct under this formulation.

Reply:
I see.